### PR TITLE
Shell: Treat builtin names as programs and suggest them

### PR DESCRIPTION
**Pull request by alimpfard**
cc @bugaevc 

I'm definitely glad I did that `ENUMERATE_SHELL_BUILTINS` thing :^)

### CODE REVIEW

File: Shell/Shell.h [hunk 1]
```diff
@@ -174,6 +174,26 @@ class Shell : public Core::Object {
 
 #undef __ENUMERATE_SHELL_BUILTIN
 
+    constexpr static const char* BuiltinNames[] = {
```
Comment (awesomekling) [inline Shell/Shell.h:177]: `builtin_names`

File: Shell/Shell.h [hunk 2]
```diff
@@ -174,6 +174,26 @@ class Shell : public Core::Object {
...
+#define __ENUMERATE_SHELL_BUILTIN(builtin) #builtin,
+
+        ENUMERATE_SHELL_BUILTINS()
+
+#undef __ENUMERATE_SHELL_BUILTIN
+    };
+    constexpr static auto BuiltinCount = sizeof(BuiltinNames) / sizeof(char*);
```
Comment (bugaevc) [inline Shell/Shell.h:184]:
```suggestion
    constexpr static auto BuiltinCount = sizeof(BuiltinNames) / sizeof(const char*);
```

😀

File: Shell/Shell.cpp [hunk 3]
```diff
@@ -1373,6 +1373,10 @@ void Shell::cache_path()
         }
     }
 
+    // add shell builtins to the cache
+    for (const auto& builtin_name : builtins())
+        cached_path.append(escape_token(builtin_name));
```
Comment (awesomekling) [inline Shell/Shell.cpp:1378]: I don't think you need the separate function for this, you can just iterate the array with range-for:

```cpp
for (auto* builtin_name : builtin_names)
    ...
```

File: Shell/Shell.h [hunk 4]
```diff
@@ -174,6 +174,26 @@ class Shell : public Core::Object {
...
+
+    static Vector<StringView, BuiltinCount> builtins()
```
Comment (bugaevc) [inline Shell/Shell.h:186]: Why even use a vector (and not say an array) if the size is statically known?

File: Shell/Shell.h (hunk 2 above)
Comment (alimpfard) [reply Shell/Shell.h:184]: but...why?
I guess...

File: Shell/Shell.h (hunk 1 above)
Comment (alimpfard) [reply Shell/Shell.h:177]: I had taken to doing PascalCase for constexpr members...sure.

File: Shell/Shell.cpp (hunk 3 above)
Comment (alimpfard) [reply Shell/Shell.cpp:1378]: range-based for works with constexpr arrays? neat!

**Pull request merged**